File: DSA/MergeSort.java
// Java code to implement Merge Sort

public class MergeSort {
	public static void main(String[] args) {
		int[] unsorted = {2, 9, 7, 1, 4, 8};
		mergeSort(unsorted, 0, 5);
		for (int i : unsorted) {
			System.out.println(i);
		}
	}

	public static void merge(int[] arr, int p, int q, int r) {
		int lengthLeft = q-p+1;
		int lengthRight = r-q;

		int[] left = new int[lengthLeft];
		int[] right = new int[lengthRight];

		for (int i=0; i<lengthLeft; i++) {
			left[i] = arr[p+i];
		}

		for (int j=0; j<lengthRight; j++) {
			right[j] = arr[q+1+j];
		}

		int i=0; 
		int j=0;
		int k=p;

		while (i<lengthLeft && j<lengthRight) {
			if (left[i]<=right[j]) {
				arr[k++] = left[i++];
			}
			else {
				arr[k++] = right[j++];
			}
		}

		while (i<lengthLeft) {
			arr[k++] = left[i++];
		}

		while (j<lengthRight) {
			arr[k++] = right[j++];
		}

	}

	public static void mergeSort(int[] arr, int p, int r) {
		if (p>=r) {
			return;
			
		}

		else {
			int q = (p+r)/2;
			mergeSort(arr, p, q);
			mergeSort(arr, q+1, r);

			merge(arr, p, q, r);
		}
	}

}

File: 100 DOC/DAY#8/mergeSort.py
# Author: Vaikan Peddi
# Date: 4th October 2023
# Description: We will be looking into a simple Divide-and-Conquer Algorithm: Merge Sort

def merge(A:list[int], p:int, q:int, r:int) -> None:
    left = A[p : q+1]
    right = A[q+1: r+1]
    
    i = 0
    j = 0
    k = p

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            A[k] = left[i]
            i += 1
        else:
            A[k] = right[j]
            j += 1
        k += 1

    if i < len(left): 
        A[k: r+1] = left[i:]
    if j < len(right): 
        A[k: r+1] = right[j:]

def merge_sort(A:list[int], p:int=0, r:int=None) -> None:
    if r == None:
        r = len(A)
    if p >= r:
        return
    q = (p + r) // 2
    merge_sort(A, p, q)
    merge_sort(A, q+1, r)
    merge(A, p, q, r)

test = [1, 6, 2, -1, -5]
merge_sort(test)
print(test)